### ISSUE TITLE
DIffusion bug when changing behaviour frequency

### ISSUE THREAD

**Issue by ryan-bournes** [BioDynaMo/biodynamo#343]
**The Problem**
When the frequency of the behavior operation is changed to something different to the diffusion operation, the diffusion operation breaks. 

**How to Replicate**
Create simple behavior where a cell consumes a fixed amount of nutrients per time step.

```
void Run(Agent* agent) override {
    if (auto* cell = dynamic_cast<Cell*>(agent)) {
            Real3 p = cell->GetPosition();
            auto* dgrid = Simulation::GetActive()->GetResourceManager()->GetDiffusionGrid( "Kalium" );
            dgrid->ChangeConcentrationBy( p, -1, InteractionMode::kAdditive);
    }
  }
```

Create a uniform substance and place a single cell at the bottom of the environment with the behavior:
```

struct FlatConc {
          real_t conc_;
          FlatConc(real_t initial_conc) { conc_ = initial_conc; }
          real_t operator()(real_t x, real_t y, real_t z) { return conc_; }
  };
  ModelInitializer::DefineSubstance(kKalium, "Kalium", 4, 0, 10);
  ModelInitializer::InitializeSubstance(kKalium, FlatConc( 1000 ));


  auto* rm = simulation.GetResourceManager();
  auto* cell = new Cell({25, 25, 0});
  cell->SetDiameter(1.0);
  cell->AddBehavior(new eat() );
  rm->AddAgent(cell);

```

Then change the behavior frequency to 2:

```
  auto* scheduler = simulation.GetScheduler();
  scheduler->GetOps("behavior")[0]->frequency_ = 2;
```

**What Happens**
As diffusion is being calculated twice per behaviour time step, we should see diffusion move twice as fast.
However what we see is that diffusion does not occur at all and also the concentration "flickers" between the voxel the cell is in and the voxel above it. 

I originally discovered this problem on an older version of BioDynaMo that I use (version 1.5.0) however I have been able to replicate the same problem on the newest version. I am using Ubuntu-22.04.

**Comment by TobiasDuswald** [comment]
Hi @ryan-bournes ,

```cpp
scheduler->GetOps("behavior")[0]->frequency_ = 2;
```
So this does not mean that diffusion is being calculated twice per timestep, this means that you compute diffusion once, then `eat()`, `eat()` again, and then diffuse again. Does this explain you observation?

If not, please collect the code in a **small** reproducer repository, zip it, and post it here. Thanks!

**Comment by ryan-bournes** [comment]
Hi @TobiasDuswald ,

The frequency is described as running the operation every nth timestep. For example, setting it to 10 runs the operation every 10th timestep. This can be found in the source code in src/core/operations/operations.h

`  
  /// Specifies how often this operation will be executed.\n
  /// 1: every timestep\n
  /// 2: every second timestep\n
  /// ...
  size_t frequency_ = 1;
`

I have attached the compressed project containing the test. An explanation for what the simulation is and how to run the test can be found at the bottom of the README that comes with the project. You can also find two animations of two different frequencies inside the output folder. 
[freq_test.zip](https://github.com/BioDynaMo/biodynamo/files/12337247/freq_test.zip)

**Comment by TobiasDuswald** [comment]
So I took a look, and I do not see any frequency-related issues. Your understanding of the `frequency_` argument is correct; however, let me emphasize again that
```cpp
scheduler->GetOps("behavior")[0]->frequency_ = 2;
```
does change the frequency of the `eat` behavior, not the frequency of the diffusion. You stated that you would expect diffusion to be twice as fast, but that is not true. You have the same total time, the same time step, and the same diffusion parameters. They all remain unaffected by the `frequency_` of `"behavior"`. The diffusion dynamics are unaffected, you only `eat` away the substance every `nth` timestep.

To avoid any issues coming from the boundary conditions (I'd recommend using `neumann` instead of `closed`), I moved the cell to `{0,0,0}` and put the domain to `[-50,50]^3`. I added print statements to the `eat` behavior displaying the current concentration, see the screenshot below. The left column shows `scheduler->GetOps("behavior")[0]->frequency_ = 1;` and the right column shows `scheduler->GetOps("behavior")[0]->frequency_ = 2;`. 

![image](https://github.com/BioDynaMo/biodynamo/assets/44771875/c2ff25f1-383c-481b-96ca-1d9d0ffab28e)

You can see that for `f=1` the behavior gets called for every time step. The green `eat` calls are missing on the right, e.g. it gets called every second time step. You can also see whenever BDM is updating the grid with the FTCS scheme, e.g., `BMD calling DiffuseWithClosedEdge ..`.

`eat` seems to work fine, e.g., takes 1 away. The diffusion is scheduled as expected and is thoroughly unit tested, so I assure you the stencil is correct. Also the scheduling of `eat` is as desired. To me, it all looks as it should.

Generally, voxels at the boundary require a different treatment compared to voxels in the inner part of the domain. Maybe the issues that you see are related to the way the boundary is treated. But you would have to understand what exactly you would like to see, generally, I'd advise having a bit of a distance between the agents and the boundary.

Rewritten code (note the diff for BDM print statement in it):
[freq_test.zip](https://github.com/BioDynaMo/biodynamo/files/12342259/freq_test.zip)

Hope that helps.

**Comment by ryan-bournes** [comment]
Sorry for the confusion, I meant that diffusion would be twice as fast relative to the eat behaviour. If you set the frequency of the behaviour to 2 and then ran for 400 timesteps, then the cell would have 'eaten' as much concentration as the simulation where frequency=1 and time=200. However because diffusion gets calculated twice between each 'eat' the actual concentration of the voxel ends up being higher in the frequency=2 simulation because the diffusion is 'faster' relative to the eat behaviour.

Going through the test that you posted, I can see that diffusion works as expected when the cell is far away (roughly 2 voxels) from the boundary, regardless of what the behaviour of the frequency is set to. The diffusion/frequency problem only occurs when a cell is in a voxel near the boundary. 

Using your suggestion, I found that using Neumann boundary conditions gives a much more realistic view of diffusion at the boundaries. My simulation relies on how cells interact with nutrients at the boundaries as it represents bacteria growing at the bottom of a closed box.

Before I begin using Neumann boundaries instead of closed boundaries, can I ask what the difference between the two types are?

Many thanks for your help.
Ryan

**Comment by romanbauer111** [comment]
Hi @ryan-bournes , to save Tobias some time:

Neumann Boundary Conditions and Closed Boundary Conditions are terms used in the context of partial differential equations and boundary value problems. They define how a mathematical model behaves at the boundaries of a region in which the equation is being solved. Let's explore the differences between these two types of boundary conditions:

Neumann Boundary Conditions:
Neumann boundary conditions specify the derivative of the solution at the boundary. In other words, they describe how the function's rate of change behaves at the boundary. Mathematically, for a function u(x) defined in a domain Ω, Neumann boundary conditions can be expressed as:
∂u/∂n = g(x),

where ∂u/∂n is the derivative of u in the direction normal to the boundary, and g(x) is a given function describing the behavior of the derivative at the boundary.

Neumann boundary conditions are often used to represent physical scenarios where quantities such as heat or mass flow across a boundary are known or controlled. For example, in heat conduction problems, Neumann boundary conditions can represent situations where heat flux is specified at the boundary.

Closed Boundary Conditions:
Closed boundary conditions, also known as Dirichlet boundary conditions, specify the value of the solution directly at the boundary. These conditions provide information about the actual function values at the boundary. Mathematically, for a function u(x) defined in a domain Ω, closed boundary conditions can be expressed as:
u(x) = f(x),

where u(x) is the function being solved, and f(x) is a given function that defines the values of u at the boundary.

Closed boundary conditions are often used to represent situations where the values of a physical quantity are fixed or controlled at the boundary. For example, in heat conduction problems, closed boundary conditions can represent scenarios where the temperature is fixed at the boundary.

In summary, the main difference between Neumann boundary conditions and closed (Dirichlet) boundary conditions lies in what they specify at the boundaries. Neumann conditions describe the behavior of the derivative of the solution at the boundary, while closed conditions specify the actual values of the solution at the boundary. The choice between these boundary conditions depends on the specific physical scenario being modeled and the information available about the system being studied.

**Comment by TobiasDuswald** [comment]
Thanks @romanbauer111!

Very small addition: in BDM you can choose the boundaries `open`, `closed`, `neumann`, and `dirichlet`. The former two are kept for backwards compatibility at the moment but I'd recommend using `neumann` or `dirichlet`. These two behave exactly as @romanbauer111 described above, i.e., setting the value of the derivative or the actual function values on the boundary, respectively.

@ryan-bournes see this old PR for more details: https://github.com/BioDynaMo/biodynamo/pull/258 . Also check the unit test that we introduced there and make sure to verify your results if your use case differs from these scenarios to be on the safe side.

**Comment by ryan-bournes** [comment]
Hi @TobiasDuswald and @romanbauer111 

Thank you both for the information! I appreciate the help.

Best regards,
Ryan
